File: suii_task_manager/cppcode/src/TaskList.cpp
#include "TaskList.h"

TaskList::TaskList()
{
    this->capacity = DEFAULT_CAPACITY;
}

TaskList::TaskList(int capacity)
{
    this->capacity = capacity;
}

TaskList::~TaskList()
{
    // free task list memory
    for (unsigned int i = 0; i < this->taskList.size(); i++)
    {
        delete this->taskList[i];
    }
}

int TaskList::indexTask(int type, int src, int dest, int obj, int container)
{
    Task t;
    t.setType(type);
    t.setSrc(src);
    t.setDest(dest);
    t.setObj(obj);
    t.setContainer(container);
    for (unsigned int i = 0; i < this->taskList.size(); i++)
    {
        if (this->taskList[i]->isEqualTo(&t))
            return i;
    }
    return -1;
}

bool TaskList::isEmpty(void)
{
    return this->taskList.size() == 0;
}

bool TaskList::isFull(void)
{
    return this->taskList.size() == this->capacity;
}

void TaskList::clearTask(void)
{
    this->taskList.clear();
}

bool TaskList::addTask (Task t)
{
    if (!this->isFull())
    {
        Task* task;
        task = new Task;
        task->setType(t.type);
        task->setSrc(t.src);
        task->setDest(t.dest);
        task->setObj(t.obj);
        task->setContainer(t.container);
        this->taskList.push_back(task);
        return true;
    }
    return false;
}

bool TaskList::addTask (Task* t)
{
    if (!this->isFull())
    {
        Task* task;
        task = new Task;
        task->setType(t->type);
        task->setSrc(t->src);
        task->setDest(t->dest);
        task->setObj(t->obj);
        task->setContainer(t->container);
        this->taskList.push_back(task);
        return true;
    }
    return false;
}

bool TaskList::removeTask(Task t)
{
    int index = indexTask(t.type, t.src, t.dest, t.obj, t.container);
    if (index == -1)
        return false;
    this->taskList.erase(this->taskList.begin() + index);
    return true;
}

bool TaskList::removeTask(Task* t)
{
    int index = indexTask(t->type, t->src, t->dest, t->obj, t->container);
    if (index == -1)
        return false;
    this->taskList.erase(this->taskList.begin() + index);
    return true;
}

bool TaskList::getNextObjToPick(Task& t, int& index)
{
    for (unsigned int i = 0; i < this->taskList.size(); i++)
    {
        if (!this->taskList[i]->picked)
        {
            t.setType(this->taskList[i]->type);
            t.setSrc(this->taskList[i]->src);
            t.setDest(this->taskList[i]->dest);
            t.setObj(this->taskList[i]->obj);
            t.setContainer(this->taskList[i]->container);
            index = i;
            return true;
        }
    }
    return false;
}

TaskList TaskList::getTasksBySrc (int src)
{
    TaskList list;
    for (unsigned int i = 0; i < this->taskList.size(); i++)
    {
        if (this->taskList[i]->src == src)
        {
            list.addTask(this->taskList[i]);
        }
    }
    return list;
}

TaskList TaskList::getTasksByDest (int dest)
{
    TaskList list;
    for (unsigned int i = 0; i < this->taskList.size(); i++)
    {
        if (this->taskList[i]->dest == dest)
        {
            list.addTask(this->taskList[i]);
        }
    }
    return list;
}

std::map<int, std::string> TaskList::getUniqueDest(void)
{
    std::map<int, std::string> uniqueDest;
    for (unsigned int i = 0; i < this->taskList.size(); i++)
    {
        uniqueDest.insert(std::make_pair(this->taskList[i]->dest, this->taskList[i]->dest_str));
    }
    return uniqueDest;
}

std::map<int, std::string> TaskList::getUniqueSrc(void)
{
    std::map<int, std::string> uniqueSrc;
    for (unsigned int i = 0; i < this->taskList.size(); i++)
    {
        uniqueSrc.insert(std::make_pair(this->taskList[i]->src, this->taskList[i]->src_str));
    }
    return uniqueSrc;
}

void TaskList::sortBySrcAndDest(void)
{
    std::sort(this->taskList.begin(), this->taskList.end(), this->sortBySourceAndDest);
}

void TaskList::printTaskList(void)
{
    for (unsigned int i = 0; i < this->taskList.size(); i++)
    {
        this->taskList[i]->printTaskData();
    }
}

File: suii_task_manager/src/suii_task_manager/task_list.py
#! /usr/bin/env python
 
from functools import cmp_to_key
from task import Task

class TaskList:
    DEFAULT_CAPACITY   = 20 

    def __init__(self):
        self.capacity   = self.DEFAULT_CAPACITY
        self.task_list  = []
    
    def get_task_index(self, task_type, source, destination, obj, container):
        task = Task()
        task.set_type(task_type)
        task.set_source(source)
        task.set_destination(destination)
        task.set_object(obj)
        task.set_container(container)
        
        for i in range(0, len(self.task_list)):
            if (self.task_list[i].is_equal_to(task)):
                return i
                
        return -1

    def is_empty(self):
        return len(self.task_list) == 0

    def is_full(self):
        return len(self.task_list) == self.capacity

    def clear_task(self):
        self.task_list = []

    def add_task (self, task):
        if not self.is_full():
            # Thanh question: can we append task directly w/o creating tmp? (memory issues?)
            tmp_task = Task()
            tmp_task.set_type(task.type)
            tmp_task.set_source(task.source)
            tmp_task.set_destination(task.destination)
            tmp_task.set_object(task.object)
            tmp_task.set_container(task.container)
            self.task_list.append(tmp_task)
            return True
        return False

    def remove_task(self, task):
        index = self.get_task_index(task.type, task.source, task.destination, task.object, task.container)
        if (index == -1):
            return False
        del self.task_list[index]
        return True

    def get_next_obj_to_pick(self):
        task = Task()
        for i in range(0, len(self.task_list)):
            if not self.task_list[i].picked:
                task.set_type(self.task_list[i].type)
                task.set_source(self.task_list[i].source)
                task.set_destination(self.task_list[i].destination)
                task.set_object(self.task_list[i].object)
                task.set_container(self.task_list[i].container)
                return True, i, task
                
        return False

    def get_tasks_by_source(self, source):
        task_list  = TaskList()
        for task in self.task_list:
            if (task.source == source):
                task_list.add_task(task)
        return task_list

    def get_tasks_by_destination(self, destination):
        task_list = TaskList()
        for task in self.task_list:
            if (task.destination == destination):
                task_list.add_task(task)
        return task_list

    def get_unique_destination(self):
        unique_dest_list = dict()
        for task in self.task_list:
            unique_dest_list[task.destination] = task.destination_str
        return unique_dest_list

    def get_unique_source(self):
        unique_src_list = dict()
        for task in self.task_list:
            unique_src_list[task.source] = task.source_str
        return unique_src_list

    def sort_by_src_and_dest(self):
        self.task_list.sort(key=cmp_to_key(self.cmp_items))

    def cmp_items(self, t1, t2):
        return (t1.source < t2.source) or (t1.source == t2.source and (t1.destination < t2.destination))
        
    def print_task_list(self):
        for task in self.task_list:
            print(task)
        # print([item.name for item in items]) ????